### PR TITLE
fix: Cancel visual mode when switching editor

### PR DESCRIPTION
**Pull request by trkoch**
Fixes https://github.com/vscode-neovim/vscode-neovim/issues/1577

### CODE REVIEW

**Comment by trkoch** [comment]
Note I "fixed" a test by changing the assertion. Judging from the title _Current mode is canceled when switching between editor panes_ I feel this makes more sense now.

**Comment by xiyaowong** [comment]
Off-topic, I think there are too many logs now, one log per step...😂

**Review by xiyaowong — CHANGES_REQUESTED** [review]

File: src/buffer_manager.ts [hunk 1]
```diff
@@ -575,6 +575,13 @@ export class BufferManager implements Disposable {
         logger.debug(`Setting active editor - viewColumn: ${activeEditor.viewColumn}, winId: ${winId}`);
         await this.main.cursorManager.updateNeovimCursorPosition(activeEditor, activeEditor.selection.active);
         try {
+            if (this.main.modeManager.isVisualMode) {
```
[the remark this answers is not in the document]
Comment (theol0403) [reply src/buffer_manager.ts:578]: I think insert should stay in insert, but I am good with canceling visual.

[the remark this answers is not in the document]
Comment (trkoch) [reply src/buffer_manager.ts:578]: @theol0403 I agree insert should stay in insert. It's intended behavior introduced with https://github.com/vscode-neovim/vscode-neovim/pull/1050, and consistent with vim.

[the remark this answers is not in the document]
Comment (trkoch) [reply src/buffer_manager.ts:578]: @xiyaowong Not sure about cmdline editing. Do you have a use case? Personally, I find `if (this.main.modeManager.isVisualMode)` more intention-revealing, but thats probably a matter of style. In any case, I suggest to keep this PR focused on fixing the referenced bug.

**Comment by trkoch** [comment]
> Off-topic, I think there are too many logs now, one log per step...😂

@xiyaowong Yeah. It's pretty crazy. It did help me though with debugging this issue.

**Review by theol0403 — APPROVED** [review]

**Review by theol0403 — APPROVED** [review]

File: src/buffer_manager.ts [hunk 2]
```diff
@@ -575,6 +575,13 @@ export class BufferManager implements Disposable {
...
+                // https://github.com/vscode-neovim/vscode-neovim/issues/1577
+                logger.debug(
+                    `Cancel visual mode to prevent selection from previous editor to carry over to active editor`,
+                );
+                await this.client.input("<Esc>");
```
Comment (justinmk) [inline src/buffer_manager.ts:583]: `:help interrupt()` may be more ideal than inputting ESC.

**Pull request merged**